Geometry/GEMGeometryBuilder/src/ME0GeometryParsFromDD.cc

```diff
@@ -8,75 +8,194 @@
-    pars.emplace_back( 0.0 ); // nStrips
-    pars.emplace_back( 0.0 ); // nPads
-
-    std::string name = fview.logicalPart().name().name();
-    std::vector<std::string> strpars;
-    strpars.emplace_back( name );
-
-    DDRotationMatrix rota = fview.rotation();
-
-    // ROOT::Math way
-    DD3Vector x, y, z;
-    rota.GetComponents( x, y, z );
-    std::vector<double> vrot(9);
-    vrot[0] = x.X();
-    vrot[1] = x.Y();
-    vrot[2] = x.Z();
-    vrot[3] = y.X();
-    vrot[4] = y.Y();
-    vrot[5] = y.Z();
-    vrot[6] = z.X();
-    vrot[7] = z.Y();
-    vrot[8] = z.Z();
- 
-    DDTranslation tran = fview.translation();
-    std::vector<double> vtra(3);
-    vtra[0] = tran.x();
-    vtra[1] = tran.y();
-    vtra[2] = tran.z();
-
-    rgeo.insert( rollDetId.rawId(), vtra, vrot, pars, strpars );
-    doSubDets = fview.nextSibling(); // go to next layer
+  LogDebug("ME0GeometryParsFromDD") << "buildLayer "<<fv.logicalPart().name().name()
+				    <<" "<< detId <<std::endl;
+  
+  DDBooleanSolid solid = (DDBooleanSolid)(fv.logicalPart().solid());
+  std::vector<double> dpar = solid.parameters(); 
+  
+  double dy = dpar[0]/cm;//length is along local Y
+  double dz = dpar[3]/cm;// thickness is long local Z
+  double dx1= dpar[4]/cm;// bottom width is along local X
+  double dx2= dpar[8]/cm;// top width is along local X
+  //dpar = solid.solidB().parameters();
+  //dz += dpar[3]/cm;// chamber thickness
+
+  ME0DetId me0id = detId.chamberId();
+  std::string name = fv.logicalPart().name().name();
+  std::vector<std::string> strpars{name};
+
+  
+  std::vector<double> pars{dx1, dx2, dy, dz};
+  std::vector<double> vtra = getTranslation(fv);
+  std::vector<double> vrot = getRotation(fv);
+
+  rgeo.insert(me0id.rawId(), vtra, vrot, pars, strpars);
+}
+
+void
+ME0GeometryParsFromDD::buildEtaPartition(DDFilteredView& fv, ME0DetId detId, RecoIdealGeometry& rgeo)
+{
+  LogDebug("ME0GeometryParsFromDD") << "buildEtaPartition "<<fv.logicalPart().name().name()
+				    <<" "<< detId <<std::endl;
+  
+  // EtaPartition specific parameter (nstrips and npads) 
+  DDValue numbOfStrips("nStrips");
+  DDValue numbOfPads("nPads");
+  std::vector<const DDsvalues_type* > specs(fv.specifics());
+  std::vector<const DDsvalues_type* >::iterator is = specs.begin();
+  double nStrips = 0., nPads = 0.;
+  for (;is != specs.end(); is++){
+    if (DDfetch( *is, numbOfStrips)) nStrips = numbOfStrips.doubles()[0];
+    if (DDfetch( *is, numbOfPads))   nPads = numbOfPads.doubles()[0];
   }
+  LogDebug("ME0GeometryParsFromDD") 
+    << ((nStrips == 0. ) ? ("No nStrips found!!") : ("Number of strips: " + std::to_string(nStrips))); 
+  LogDebug("ME0GeometryParsFromDD") 
+    << ((nPads == 0. ) ? ("No nPads found!!") : ("Number of pads: " + std::to_string(nPads)));
+  
+  // EtaPartition specific parameter (size) 
+  std::vector<double> dpar = fv.logicalPart().solid().parameters();
+
+  double dy = dpar[0]/cm;//length is along local Y
```

Comment:
the repeated code in this PR is quite substantial.Consider defining a few helper functions to remove this duplication.